MGMT-9722 - Redesign the review and create step

jgyselov opened the pull request:
https://issues.redhat.com/browse/MGMT-9722

KNI tests: https://gitlab.cee.redhat.com/ocp-edge-qe/kni-assisted-installer-auto/-/merge_requests/966

AI lib tests: https://github.com/openshift-assisted/openshift-assisted-ui/pull/25

**Broad overview:**
![Screenshot from 2022-12-19 11-43-13](https://user-images.githubusercontent.com/87187179/208409299-ffc04ce2-ea8a-4978-b63c-af02e41bd886.png)

**Expanded validations**
![Screenshot from 2022-12-19 11-43-47](https://user-images.githubusercontent.com/87187179/208409513-710592f3-476a-4d00-b0a3-b4b199a46d34.png)

**Failing validations:**
![Screenshot from 2022-12-19 11-46-26](https://user-images.githubusercontent.com/87187179/208409609-92e4a806-6b5b-40e3-8d21-7f90d708021a.png)

![Screenshot from 2022-12-19 11-46-39](https://user-images.githubusercontent.com/87187179/208409602-d9d56e2b-3b3f-4351-ba54-70e1fe3408c7.png)

**Soft validations:**
![Screenshot from 2022-12-19 11-48-07](https://user-images.githubusercontent.com/87187179/208409797-736d087f-a488-433f-9932-389b0859ce0a.png)
![Screenshot from 2022-12-19 11-47-55](https://user-images.githubusercontent.com/87187179/208409746-cc9580b8-7c87-4253-bfd1-a34cb32743b1.png)

jgyselov commented:
**Updated:**

- "Operators" (previous - wrong - title "Storage") section only appears when enabled.
- The same applies for disk encryption.
- API/Ingress VIP changed to API/Ingress IP to stay consistent with the Networking page

**With Operators section and disc encryption:**
![image](https://user-images.githubusercontent.com/87187179/208678211-2b38292b-8bdb-4a68-84d7-1625c2ef6cf9.png)

**Without Operators section:**
![image](https://user-images.githubusercontent.com/87187179/208678321-948bbbd7-2696-4c4a-bffc-c0f0555dcd0d.png)

**With disk encryption:**
![image](https://user-images.githubusercontent.com/87187179/208681164-4d734ad6-67c7-444a-8158-480192f64b26.png)

**Networking step:**
![image](https://user-images.githubusercontent.com/87187179/208681638-ffa287d3-06ab-45cc-8694-3b2775e9f2c1.png)

celdrake commented on src/ocm/components/clusterDetail/ClusterProperties.tsx:
```diff
@@ -37,23 +37,26 @@ type ClusterPropertiesProps = {
 };
 
-const getDiskEncryptionEnabledOnStatus = (diskEncryption: DiskEncryption['enableOn']) => {
+export const getStackType = (cluster: Cluster): string =>
```
```suggestion
export const getStackTypeLabel = (cluster: Cluster): string =>
```

celdrake commented on src/ocm/components/clusterDetail/ClusterProperties.tsx:
```diff
@@ -37,23 +37,26 @@ type ClusterPropertiesProps = {
+export const getStackType = (cluster: Cluster): string =>
+  isDualStack(cluster) ? 'Dual-stack' : 'IPv4';
+
+export const getDiskEncryptionEnabledOnStatus = (diskEncryption: DiskEncryption['enableOn']) => {
```
This looks like it could be a React Component instead of a function. Feel free to ignore :)

jgyselov replied:
Keeping this one as is.

jgyselov commented:
@celdrake  @ammont82 FIY, I have fixed a good number of Prettier warnings but because the number in the master branch keeps changing all the time, I've removed the changes done to `package.json` to avoid constant conflicts. I'll be happy to update that file once we are ready to merge.

celdrake commented on src/ocm/components/clusterConfiguration/review/ReviewValidations.tsx:
```diff
@@ -0,0 +1,191 @@
+import { ValidationsInfo as ClusterValidationsInfo } from '../../../../common/types/clusters';
+import { ValidationsInfo as HostValidationsInfo } from '../../../../common/types/hosts';
+
+const ValidationsDetailExpanded = ({ cluster }: { cluster: Cluster }) => {
```
Could you please add the label for this validation?
`time-synced-between-host-and-service check failed`

![new-validation](https://user-images.githubusercontent.com/829045/208869327-eb9da61a-9f1d-41e7-93ba-152a3fbb8d8d.png)

celdrake replied:
Personally I would prefer if validations were grouped by severity, eg. all host validations errors together, etc but this probably wasn't on the design.

ammont82 replied:
I think the same.

jgyselov replied:
Labels added.

celdrake reviewed (changes requested):
@jgyselov Looks very nice!

I noticed however that in my screen the font size for the block below seems smaller, whereas on your screnshots they look the same.
![font-sizes](https://user-images.githubusercontent.com/829045/208870463-fda97cec-945a-45f0-adad-6880c665eefa.png)

ammont82 commented:
> @jgyselov Looks very nice!
> 
> I noticed however that in my screen the font size for the block below seems smaller, whereas on your screnshots they look the same. ![font-sizes](https://user-images.githubusercontent.com/829045/208870463-fda97cec-945a-45f0-adad-6880c665eefa.png)

I have the same problem in my screen:
![image](https://user-images.githubusercontent.com/11390125/208872131-9c95b8d4-435b-4878-88c6-01c7541f0020.png)

jgyselov commented on src/ocm/components/clusterConfiguration/review/ReviewSummary.tsx:
```diff
@@ -0,0 +1,330 @@
+  );
+};
+
+const ReviewOperatorsTable = ({ cluster }: { cluster: Cluster }) => {
```
It was an easy fix, LVM added. It will only show for SNO clusters instead of ODF.

![image](https://user-images.githubusercontent.com/87187179/208883347-a010f3f5-7ee3-4fbb-a0ae-d5a486fdfe6a.png)

celdrake commented on src/common/api/swagger.yaml:
```diff
@@ -6242,10 +6243,10 @@ definitions:
-      - 'api-vip-valid'
-      - 'ingress-vip-defined'
-      - 'ingress-vip-valid'
+      - 'api-vips-defined'
```
Last thing I knew was that the API changes for Dual Stack VIPs wouldn't be released on next version.
Maybe we could update the types to have both IDs, with a TODO to remove the current one, but then we're safe as to when the API change happens.

nirfarkas commented:
Looks great!
One comment:
In the "Operators" section when an operator was not selected, we shouldn't show it at all IMO.

jgyselov commented on src/ocm/components/clusterConfiguration/review/ReviewSummary.tsx:
```diff
@@ -0,0 +1,330 @@
+};
+
+export const ReviewSummary = ({ cluster }: { cluster: Cluster }) => {
+  const showOperatorsSummary =
```
Keeping this _as is_. The check is for the three operators that we expose in the Operators step. If the condition gets more complicated one day we can revisit it.

jgyselov commented:
> Looks great! One comment: In the "Operators" section when an operator was not selected, we shouldn't show it at all IMO.

We aren't. :slightly_smiling_face:

ammont82 reviewed (approved):
Good job @jgyselov !

jkilzi commented on src/ocm/components/clusterConfiguration/review/ReviewOperatorsTable.tsx:
```diff
@@ -0,0 +1,67 @@
+            title: hasEnabledOperators(cluster.monitoredOperators, OPERATOR_NAME_CNV)
+              ? 'Enabled'
+              : 'Disabled',
+            props: { 'data-testId': 'openshift-virtualization' },
```
@jgyselov @celdrake this should be 'data-testid' all in lowercase
It is all over the place, and produces the following error in the console:
![image](https://user-images.githubusercontent.com/77008341/209287749-ec3e528d-d8e2-474b-9e84-99fd327a2279.png)

celdrake replied:
Well spotted, added a PR to fix it:
https://github.com/openshift-assisted/assisted-ui-lib/pull/1832

jkilzi commented on src/ocm/components/clusterConfiguration/review/ReviewClusterDetailTable.tsx:
```diff
@@ -0,0 +1,75 @@
+  }, [cluster, infraEnv?.staticNetworkConfig]);
+
+  return (
+    <Table
```
@jgyselov 
**The tables** also need an `aria-label` attr. This could be something like `aria-label="cluster-details-review-table"`
